osquery/tables/system/npm_packages.cpp

```diff
@@ -0,0 +1,178 @@
+
+#include <stdlib.h>
+#include <string>
+
+#include <osquery/core/tables.h>
+#include <osquery/filesystem/filesystem.h>
+#include <osquery/logger/logger.h>
+#include <osquery/utils/info/platform_type.h>
+#include <osquery/utils/json/json.h>
+#include <osquery/worker/ipc/platform_table_container_ipc.h>
+#include <osquery/worker/logging/glog/glog_logger.h>
+
+#ifdef WIN32
+#include "windows/registry.h"
+#endif
+
+namespace fs = boost::filesystem;
+
+namespace osquery {
+namespace tables {
+
+const std::set<std::string> kNodeModulesPath = {
+    "/usr/local/lib",
+    "/opt/homebrew/lib",
+    "/usr/lib",
+    "/home/%/.npm-global/lib",
+    "/Users/%/.npm-global/lib",
+    "C:\\Users\\%\\AppData\\Roaming\\npm"};
+
+const std::vector<std::string> kPackageKeys{
+    "name", "version", "description", "homepage"};
+
+const std::string kWinNodeInstallKey = "SOFTWARE\\Node.js\\InstallPath";
+
+void genNodePackage(const std::string& file, Row& r, Logger& logger) {
+  std::string json;
+
+  if (!readFile(file, json).ok()) {
+    logger.vlog(1, "Could not read package.json from: " + file);
+    return;
+  }
+
+  auto doc = JSON::newObject();
+  if (!doc.fromString(json) || !doc.doc().IsObject()) {
+    logger.vlog(1, "Could not parse package.json from: " + file);
+    return;
+  }
+
+  // create row from json and add it to results
+  for (const auto& key : kPackageKeys) {
+    if (doc.doc().HasMember(key)) {
+      const auto& value = doc.doc()[key];
+      r[key] = (value.IsString()) ? value.GetString() : "";
+    }
+  }
+
+  // Manually get nested key (Author name)
+  if (doc.doc().HasMember("author")) {
+    const auto& author = doc.doc()["author"];
+    if (author.IsString()) {
+      r["author"] = author.GetString();
+    } else if (author.IsObject()) {
+      if (author.HasMember("name")) {
+        const auto& author_name = author["name"];
+        r["author"] = (author_name.IsString()) ? author_name.GetString() : "";
+      }
+    }
+  }
+
+  // Manually get license to support deprecated licence schema.
+  // In the current schema it is a string, but in previous versions it is a
+  // dictionary with url and type
+  if (doc.doc().HasMember("license")) {
+    const auto& license = doc.doc()["license"];
+    if (license.IsString()) {
+      // Current license schema is just a top level string
+      r["license"] = license.GetString();
+    } else {
+      // If its not a string, is it a dict with 'url' ?
+      if (license.HasMember("url")) {
+        const auto& license_url = license["url"];
+        if (license_url.IsString()) {
+          // Fallback to displaying deprecated licence url
+          r["license"] = license_url.GetString();
+        }
+      }
+    }
+  }
+}
+
+void genNodeSiteDirectories(const std::string& site,
+                            QueryData& results,
+                            Logger& logger) {
+  std::vector<std::string> manifest_paths;
+  resolveFilePattern(site + "/node_modules/%/package.json", manifest_paths);
```

Comment:
consider initializing 2nd path as a boost::filesystem::path and then join both paths like that:
`boost::filesystem::path p("/node_modules/%/package.json")
resolveFilePattern(site / p, manifest_paths);`

Reply:
Good suggestion. Added!